RocketChat/Rocket.Chat.ReactNative
# Chore: Start Typescript migration

## Proposed changes
The purpose of this PR is to start the project's migration to typescript.
With that, we need to pay attention to the following points:

- Add a `tsconfig.json` file. In this process, I created this file with the default rules. At the moment, I'm allowing the `type: any` to be used, as this is the beginning of the migration, we still don't know all the types used, in the future it will be necessary to disable this option.
- Update eslint config to support .ts and .tsx files. In this process, was necessary override some js rules, so I created separated rules to .ts. I also changed our extends rules from `airbnb` to `@rocket.chat/eslint-config`, since our company has default lint rules, that's no reason to not use it.
- We have adopted `interfaces` as the default for typing in our repository, following the example below:

```
interface IActionSheetData {
	options: any;
	headerHeight?: number;
	hasCancel?: boolean;
	customHeader: any;
}
```
All interface name needs start with an `I`, as `IActionSheetData `, `IAvatar`, `IButton`... This is a common pattern, and helps all to know when see an interface on the codebase.

- Componentes migrated:
  - `app/constants`;
  - `app/containers`;
  - `app/presentation`;
  - In this process, was necessary migrated some separate files, for example: `app/lib/Navigation.ts` and `app/lib/ShareNavigation.ts`. To use your resources elsewhere.


### Prettier
Taking advantage of all the changes made on `lint`, I also set up `prettier` in the project, to help us with te code formatting.
This tool has very interesting features, and I recommend that all developers involved in this project configure their IDE for automatic code checking and integration.

To more information, please see the link: [Prettier](https://prettier.io/docs/en/index.html).

Thinking about ensuring the quality of the code, I took the liberty of changing our pre-commit script, to not only run lint, but also check the formatting of the code using prettier. The new script now runs this command: 

`"prettier-lint": "yarn lint --fix && prettier --write ."`

## Important
Depends on https://github.com/RocketChat/Rocket.Chat.js.SDK/pull/145
Depends on https://github.com/RocketChat/react-native-ui-lib/pull/3

## Issue(s)	
Closes #786

## How to test or reproduce


We can test the `prettier` following this steps:
To check your code formatting, run this on your terminal:

```sh
yarn prettier-check
```

To fix your code formatting issues, run this on your terminal:

```sh
yarn prettier
```

## Screenshots
N/A

## Types of changes




## Checklist



## Further comments
